Updates to the library for better maintainability

marz619 opened the pull request:
This is a rework (redux) of the library to make it easier to follow the structure and apply some DRY to the codebase.
* It extracts the -repeated equality and copy utility functions that were previously inlined- into their own separte file/functions. See [`utils.go`](https://github.com/gadventures/apidoc/pull/1/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915)
* rename `serDataType` to `encodeType`, updates the constant names, as well as the corresponding `encode*` & `decode*` functions to have consistent naming scheme.
* removes `glide` and the old `vendor` style of depedency management and adds `go.mod`
* Extracts the `Etag` type and methods into `etag.go`
* renames files to not require the `document_` prefix

---

### Future TODOs:
* pre-commit
* more tests
* rename `master` branch to `main`
* CI checks & tests - maybe using Github actions
* Consider adding a `Diff` utility

marz619 committed:
Add go.mod, go.sum and remove glide.yaml

marz619 committed:
move type Etag to its own file

marz619 committed:
govet / golint fixes

marz619 committed:
remove vendor from .gitignore

marz619 committed:
Refactoring the code into separate utils files

marz619 commented on document.go:
```diff
@@ -7,93 +7,51 @@ import (
-Document for nested objects
-nil for nils
-*/
+// Document represents the a single G API resource. Internaly its contents
```
```suggestion
// Document represents a single G API resource. Internally its contents
```

marz619 commented on document.go:
```diff
@@ -7,93 +7,51 @@ import (
 
+// Keys is a convenience method to return the Keys in this Document, optionally
+// sorting them.
+func (d Document) Keys(sorted bool) []string {
```
Separate this into `Keys` and `KeysSorted`

marz619 commented on document.go:
```diff
@@ -7,93 +7,51 @@ import (
 	return make(Document)
 }
 
+// Keys is a convenience method to return the Keys in this Document, optionally
+// sorting them.
+func (d Document) Keys(sorted bool) []string {
+	keys := make([]string, 0, len(d))
+	for k := range d {
+		keys = append(keys, k)
+	}
+	if sorted {
+		sort.Strings(keys)
+	}
+	return keys
+}
+
```
```suggestion
// Keys is a convenience method to return the keys in this Document
func (d Document) Keys() []string {
	keys := make([]string, 0, len(d))
	for k := range d {
		keys = append(keys, k)
	}
	return keys
}

// KeysSorted is a convenience method to return the keys in this Document, in sorted
// order
func (d Document) KeysSorted() []string {
	keys := d.Keys()
	sort.Strings(keys)
	return keys
}

```

marz619 commented on document.go:
```diff
@@ -106,11 +68,11 @@ func (d Document) Equal(other Document) bool {
 	sort.Strings(sortedKeys)
 	for _, k := range sortedKeys {
```
```suggestion
	// TODO(remove above)
	for _, k := range d.KeysSorted() {
```

marz619 commented on document.go:
```diff
@@ -122,54 +85,35 @@ func (d *Document) Copy() *Document {
 
-// TraverseFunc is an alias for function that will be executed
-// for every node in Document tree
+// TraverseFunc is an alias for function that will be executed for every item
+// in the Doucument
 type TraverseFunc func(*Document, string, interface{})
```
```suggestion
// TraverseFunc is an alias for a function that will be executed for every item
// in the Document
type TraverseFunc = func(*Document, string, interface{})
```

marz619 commented on document.go:
```diff
@@ -122,54 +85,35 @@ func (d *Document) Copy() *Document {
-// TraverseCall will visit every node in Document tree
-// and call provided f on each node
+// TraverseCall will visit every item in Document and call the provided
+// traversel func on each item
```
```suggestion
// TraverseFunc on each item
```

marz619 commented on document_binencode.go:
```diff
@@ -67,125 +91,114 @@ func encodeDocument(w io.Writer, doc Document, sortKeys bool) error {
+		return err
+	}
+	data := []byte(str)
+	err = binary.Write(w, binary.LittleEndian, int64(len(data)))
```
CHECK: is it correct to use `len` here - what is its behaviour in returning the length of a string? does it return the length of bytes, or length of runes?

marz619 replied:
Yes it is, because we are coercing the string to a byte-slice.

marz619 commented on document_binencode.go:
```diff
@@ -67,125 +91,114 @@ func encodeDocument(w io.Writer, doc Document, sortKeys bool) error {
 // decode here
 
-func decodeSerType(r io.Reader) (serDataType, error) {
+func decodeSerType(r io.Reader) (serializeDataType, error) {
```
rename to `decodeSDT` for consistency

marz619 commented on go.mod:
```diff
@@ -0,0 +1,5 @@
+module github.com/gadventures/apidoc
+
+go 1.17
```
Update to go 1.19

marz619 commented on json.go:
```diff
@@ -103,34 +102,41 @@ func (d Document) WriteOutJSON(w io.Writer) error {
-	w.WriteRune('{')
-	for idx, key := range keys {
+
+	for idx, key := range doc.Keys(true) {
```
```suggestion
	for idx, key := range doc.KeysSorted() {
```

marz619 committed:
update go.mod from 1.17 to 1.19

marz619 committed:
delete departure_test.go and move tests into fixtures_test.go

marz619 committed:
typo, formatting fixes

marz619 commented on utils.go:
```diff
@@ -0,0 +1,63 @@
+
+// equality utils
+
+func docEqual(k string, d Document, o interface{}) bool {
```
rename these equality methods to end with `s`?
* `docEqual` -> `docEquals`
*  `sliceEqual` -> `sliceEquals`
*  `valEqual` -> `valEquals`

marz619 commented on serialize.go:
```diff
@@ -2,49 +2,81 @@ package apidoc
+}
+
+func encodeSDT(w io.Writer, t serializeDataType) error {
+	var ary [1]byte = [1]byte{byte(t)}
```
extract this into a method on the type `serializeDataType`

marz619 committed:
rename serialize.go to encode.go

Add the Document.Ary method

marz619 committed:
rename utils.go equality methods

marz619 committed:
rename encodeEDT and decodeEDT to include [en/de]codeEncodeType

rafikdraoui commented on document.go:
```diff
@@ -7,110 +7,67 @@ import (
 	data, err := json.MarshalIndent(d, "", "  ")
 	if err != nil {
-		return err.Error()
+		panic(err)
```
```suggestion
		// Using `%#v` flag to avoid recursive calls to String
		return fmt.Sprintf("%#v", d)
```

Better not to panic.

rafikdraoui commented on document.go:
```diff
@@ -7,110 +7,67 @@ import (
 
-// Equal compares if two documents are the same
+// Equal compares if two Documents are the same
 func (d Document) Equal(other Document) bool {
```
Since `Document` is just a map under the hood, wouldn't `reflect.DeepEqual` be sufficient here? The tests still pass with this change in:


```go
func (d Document) Equal(other Document) bool {
	return reflect.DeepEqual(d, other)
}
```

From https://pkg.go.dev/reflect#DeepEqual:

```
Map values are deeply equal when all of the following are true: they are
both nil or both non-nil, they have the same length, and either they are
the same map object or their corresponding keys (matched using Go
equality) map to deeply equal values.
```

rafikdraoui commented on document.go:
```diff
@@ -187,8 +125,7 @@ func (d *Document) UnmarshalBinary(data []byte) error {
 	if err != nil {
-		return fmt.Errorf(
-			"binary.Read of checksum failed: unmarshalling: %v", err)
+		return fmt.Errorf("binary.Read of checksum failed: unmarshaling: %v", err)
```
```suggestion
		return fmt.Errorf("binary.Read of checksum failed: unmarshaling: %w", err)
```

rafikdraoui commented on document.go:
```diff
@@ -229,63 +165,39 @@ func (d Document) MarshalBinary() ([]byte, error) {
+			ErrorID:        errID,
+			HTTPStatusCode: status,
+			Message:        msg,
+			URI:            uri,
```
```suggestion
	_, hasErr := d["error_id"]
	if !hasErr {
		return nil
	}
	errID, _ := d["error_id"].(string)
	msg, _ := d["message"].(string)
	status, _ := d["http_status_code"].(int)
	return &ErrGAPI{
		ErrorID:        errID,
		HTTPStatusCode: status,
		Message:        msg,
		URI:            uri,
	}
```

I believe returning early if it's not the "happy path" is a bit more readable.

rafikdraoui commented on encode.go:
```diff
@@ -2,49 +2,87 @@ package apidoc
-	var ary [1]byte
-	ary[0] = byte(typ)
-	if l, err := w.Write(ary[:]); l != 1 {
+// Ary return the ary byte value of the encodeType
```
What does "ary" mean here? "Array"?

marz619 replied:
I think related to `arity` ... e.g. n-ary - to correspond to the value of the `encodeType`. But maybe something like `BinaryValue()` might be more appropriate, or `ByteValue()`.

rafikdraoui commented on encode.go:
```diff
@@ -2,49 +2,87 @@ package apidoc
+	case encodeTypeNil:
+		return "Nil"
+	default:
+		panic(fmt.Sprintf("unrecognized encodeType %d %T %#v", e, e, e))
```
```suggestion
		return fmt.Sprintf("Unknown(%d)", e)
```

rafikdraoui commented on encode.go:
```diff
@@ -297,16 +324,16 @@ Loop:
-		case serDBool:
-		case serDFloat64:
-		case serDNil:
+		case encodeTypeDocumentStart:
+			val, _ = decodeDocument(r)
+		case encodeTypeListStart:
+			val, _ = decodeList(r)
```
I think we can keep `err` here, and then add the following just after the switch statement:

```go
		if err != nil {
			return lst, err
		}
```

marz619 replied:
restructured the loop a little bit and removed the `break Loop` as well - I've always tried to avoid Labels in code.

rafikdraoui commented on encode.go:
```diff
@@ -2,49 +2,87 @@ package apidoc
+
 func encodeDocument(w io.Writer, doc Document, sortKeys bool) error {
-	if err := encodeSerType(w, serDDocumentStart); err != nil {
+	err := encodeEncodeType(w, encodeTypeDocumentStart)
+	if err != nil {
```
```suggestion
	if err := encodeEncodeType(w, encodeTypeDocumentStart); err != nil {
```

rafikdraoui commented on json.go:
```diff
@@ -103,34 +102,41 @@ func (d Document) WriteOutJSON(w io.Writer) error {
 		if err != nil {
 			return err
 		}
 		w.WriteRune(':')
```
```suggestion
		_, err = w.WriteRune(':')
		if err != nil {
			return err
		}
```

We could instead have:
```go
if _, err := w.WriteRune(r); err != nil {
	return err
}
```

but if we do that, we might want to change the other calls to WriteRune in this function as well for consistency.

Another idea to reduce verbosity throughout this file would be to have a custom writer `w` that saves its first error and stops writing if `w.err != nil`, similar to `bufio.Scanner`. That way, we would only need to check errors at the end of the loop.

i.e. 
```
w.WriteRune(...)
...
w.WriteRune(...)
...
if err := w.Err(); err != nil {
	return err
}
```

But I don't think it's necessarily worth the code churn, at least not until we believe that this package will see more code changes in the future (and in any case it shouldn't be part of this PR).

marz619 replied:
This is a WIP, but will be deferred to a different PR

marz619 replied:
Something like this?
```go
// RuneWriter interface ensures the WriteRune method is available
type RuneWriter interface {
	WriteRune(rune) (int, error)
}

// multiErrRuneWriter allows a caller to use this struct multiple times
// and catch or ask for an error at the end either through defer-ing the Err()
// call, or just returning rw.Err()
type multiErrRuneWriter struct {
	writer RuneWriter
	err    error
}

func (w *multiErrRuneWriter) WriteRune(r rune) error {
	// allow calls to "succeed" if we already have an error
	if w.err != nil {
		return nil
	}
	_, err := w.writer.WriteRune(r)
	if err != nil {
		w.err = err
		return nil
	}
	return nil
}

func (w *multiErrRuneWriter) Err() error {
	return w.err
}
```

OR - `WriteRune` without any errors
```go
func (w *multiErrRuneWriter) WriteRune(r rune) {
	// allow calls to "succeed" if we already have an error
	if w.err != nil {
		return
	}
	_, w.err = w.writer.WriteRune(r)
	return
}

rafikdraoui replied:
Yes, except that WriteRune doesn't need to return anything. Here's an example of this pattern: https://go.dev/blog/errors-are-values

marz619 replied:
haha - yea, I had that same thought after I posted the initial comment, and quickly edited it with a followup that coincidentally looks almost exactly like the example in the blog post 😄

rafikdraoui commented on json.go:
```diff
@@ -206,12 +218,14 @@ func jsonMarshalFloat64(w *bufio.Writer, n float64) error {
+	return err
 }
 
 func jsonMarshalList(w *bufio.Writer, list []interface{}) error {
```
We should check the errors on `WriteRune` in this function and `jsonMarshalNil` (they are not part of the diff, so I can't make inline comments).

rafikdraoui commented on utils.go:
```diff
@@ -0,0 +1,63 @@
+	}
+}
+
+// equality utils
```
These can be removed if we decide to make `Document.Equal` be `reflect.DeepEqual`.

rafikdraoui commented on fixtures_test.go:
```diff
@@ -1,403 +1,99 @@
+	}
+}
+
+var departureBlob = `{"id":"733048","href":"https://rest.gadventures.com/departures/733048","date_created":"2016-05-09T14:53:24Z","date_last_modified":"2017-02-08T17:59:06Z","name":"Delta & Falls Overland (Westbound)","start_date":"2017-04-29","finish_date":"2017-05-07","product_line":"DZFO","sku":"GPFDZFO170429-O1","flags":[],"start_address":{"street":"Stand 1385","city":"Victoria Falls","countr ...
```
An alternative would be to put this in a JSON file withing a `testdata` directory, and have the tests read from it instead of inlining here. I don't have strong opinions either way.

marz619 replied:
Yea - I was considering doing that when I was collapsing the blob into a single line.

rafikdraoui reviewed (commented):
For what it's worth, I tried this branch on the only project I know of which uses this module, and its tests passed 👍 

Regarding the TODOs in the PR description, they could be done incrementally in their own future PRs. This one is already pretty big as it is.

ammaar-gadv committed:
use %w to allow error unwrapping

c.f. https://pkg.go.dev/fmt#Errorf

Co-authored-by: Rafik Draoui <rafik@rafik.ca>

marz619 committed:
move fixture blob data into testdata/ folder

marz619 committed:
use reflect.DeepEqual for Document.Equal

marz619 committed:
rename Document.Ary to ByteValue

marz619 committed:
prefer happy in Document.GAPIError when returning ErrGAPI struct

marz619 committed:
avoid panics in favour of returning concrete values

marz619 committed:
restructure code to use if err := ...; err != nil structure

This is more idiomatic with Go, and also ensures that we can carry the
err through the function and return it where appropriate

marz619 committed:
capture all err's when calling WriteRune

rafikdraoui commented on fixtures_test.go:
```diff
@@ -1,403 +1,131 @@
+	"testing"
+)
+
+func loadTestData(filename string) (string, error) {
+	file, err := os.Open(fmt.Sprintf("./testdata/%s", filename))
+	if err != nil {
+		return "", fmt.Errorf("failed to load file %s - %s", filename, err.Error())
+	}
+	defer file.Close()
+	blob, err := io.ReadAll(file)
+	if err != nil {
+		return "", fmt.Errorf("failed to read file %s - %s", filename, err.Error())
+	}
+	return string(blob), nil
+}
+
+// hold the data in memory
+var departureBlob, errorBlob string
+
+func setup() {
+	var err error
+	departureBlob, err = loadTestData("departure.json")
+	if err != nil {
+		panic(err)
+	}
+	errorBlob, err = loadTestData("gapi_error.json")
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestMain(m *testing.M) {
+	setup()
+	os.Exit(m.Run())
+}
```
```suggestion
func loadTestData(t *testing.T, name string) string {
	t.Helper()
	data, err := os.ReadFile(fmt.Sprintf("testdata/%s", name))
	if err != nil {
		t.Fatal(err)
	}
	return string(data)
}
```

and then the tests can call `data := loadTestData(t, "fname.json")` directly. By using `os.ReadFile`, we can consolidate the opening and the reading of the file into a single step. By using `t.Helper()`, we can make it easier to use in tests, while still getting useful error messages.

Actually, since the string data is always wrapped into a `bytes.NewBufferString` in the tests, `loadTestData` could return `[]bytes` directly, and then the tests could use ` bytes.NewReader` instead.

rafikdraoui commented on encode.go:
```diff
@@ -0,0 +1,347 @@
+)
+
+// ByteValue return the byte value of the encodeType
+func (e encodeType) ByteValue() []byte {
```
Renaming from `Ary` to `ByteValue` is a breaking change, but this is fine since the library is still pre-1.0. And that said, I'm pretty confident that no one is using that function, given that the type itself is not exported, and it's not a return type of any of the exported functions.

Since it's a breaking change already, maybe it makes sense to rename it to `byteValue`, and leave it as an internal helper function?

marz619 replied:
There is not an `Ary` function currently in the mainline, so this is an entirely new add, but agree with making it an internal helper method 👍🏽

rafikdraoui reviewed (approved):
Two comments, but otherwise looks good to me.

marz619 committed:
make encodeType.byteValue an internal method

marz619 committed:
simplify fixture test file loading per PR review

c.f. https://github.com/gadventures/apidoc/pull/1#discussion_r1042356582
